tempesta_fw/http.c

```diff
@@ -1328,6 +1328,38 @@ tfw_http_req_destruct(void *msg)
 		tfw_http_sess_put(req->sess);
 }
 
+/*
+ * Request messages that were forwarded to a backend server are added
+ * to and kept in @fwd_queue of the connection @conn for that server.
+ * If a paired request is not found, then the response must be deleted.
+ *
+ * If a paired client request is missing, then it seems upsream server
+ * is misbehaving, so the caller has to drop the server connection.
+ */
+static int
+tfw_http_popreq(TfwHttpMsg *hmresp)
+{
+	TfwHttpReq *req;
+	TfwSrvConn *srv_conn = (TfwSrvConn *)hmresp->conn;
+	struct list_head *fwd_queue = &srv_conn->fwd_queue;
+
+	spin_lock(&srv_conn->fwd_qlock);
+	req = list_first_entry_or_null(fwd_queue, TfwHttpReq, fwd_list);
+	while (req && req->pair)
+		req = list_next_entry(req, fwd_list);
```

Comment:
I feel a little uneasy about this search loop. I think the code should be made more robust. The concerns are:

1. It disregards `msg_sent` which serves the role of a stop guard for the list of sent messages. If there's a bug somewhere (req/resp get unsynchronized), this loop may cross it and go over unsent messages.
2. It disregards the end of the queue as well. It may get the queue head which is not a `TfwHttpReq{}` structure.
3. There's no case where `req` is not found (see [2] above). The latest `req` that was looked up will be paired with the response.

All of these are situations that should not happen in the course of normal work. That means, one or more of them will definitely happen, sooner or later. :-)